src/alarms/alarm-handling.md

```diff
@@ -0,0 +1,102 @@
+| type-string       | Device-unique type string (kind of alarm ID or alarm name), used to reference the once occured alarm again, e.g. in case of updates (e.g. "temperature_sensor_loss") |
+| message-string    | Human readable short information about alarm reason (e.g. "Temperature sensor does not respond") |
+| severity-string   | Could be "CRITICAL", "MAJOR", "MINOR" or "WARNING" |
+| status-string     | Could be "ACTIVE" or "CLEARED"<br/>
+| time-string       | Timestamp indicating when the alarm (or the alarm-update) had occured (in [ISO 8601 format](http://www.w3.org/TR/NOTE-datetime#)) |
+
+# Data flow
+
+Figure below illustrates the data flow from Customer Application broker/thin-edge up to cloud. It shows especially the use of MQTT QOS=1, MQTT Retain and Mosquitto Persistence to achieve reliability alarm state-change transfer.
+
+![Sequence Diagram Update SW-list](images/alarm_dataflow.svg)
+
+**To be decided:** Interface from Mapper to Cloud to be defined. Two options are possible:
+1) **JSONviaMQTT:** To have reliability use QOS=1. Open issue: Not yet completely confirmed that C8Y sends PUBACK when msg was processed, instead just on arrival.
+2) **HTTP REST:** HTTP response indicates if message was processed sucessfully. Anyway HTTP REST would increase complexity of implementation and increase data traffic.
+
+## Requirements
+|   |
+|---------------|
+| Thin-edge shall transfer most recent state of an alarm to the cloud. |
+| Other local components (as local processes or connected clients) can consume alarm-states from thin-edge by subscribing local broker topics. A local consumer can be made to process full alarm-state history by using "Clean Start Flag=0", a constant "client Id" and "QOS>0". |
+| The same alarm-state message shall be transfered just once to the cloud.<br/>An alarm-state message is treated as different when at least one field (see section 'Information Set per Alarm' above) in the message differs to the message before.<br/>To avoid duplicates some mapper-specific topic shall be used (retained) to store and identify messages that were already transfered to the cloud (e.g. `mapper/c8y/ack/alarms/<alarm>`). |
+
+
+# Public MQTT-based alarm interface
+
+Similar to thin-edge'S measurement interface, alarm interface is based on MQTT topics.
+
+## Topic structure and payload
+
+Proposal to have "severity" and "type-string" as topics:
+
+**thin-edge JSON (alarm) format:**
+```
+Topic:   tedge/alarms/<severity-string>/<type-string>
+Payload: {
+  "message":  <message-string>,
+  "status":   <status-string>,  
+  "time":     <time-string>
+}
+```
+
+NOTE: The sub-topic `childs` can not be used as type-string, since it is reseved for child-device addressing (see below).
+
+**Adressing Child-Devices:**
+To address an alarm to a child-device the sub-topic `childs` followed by the child's device id has to be used as below:
+```
+Child-Device Topics:   tedge/alarms/childs/<child-device id>/<severity-string>/<type-string>

```

Comment:
I'm not really happy with this `childs` prefix, mainly because this is not consistent with what we have for `measurements`.

Reply:
You are right. But our plan is to change the topic structure for "measurements" to become consistent with "alarms".

Finally for child-devices addressing I wish to have a same topic-scheme for all our "thin-edge" existing and upcoming topics. Something like:

```tedge/<API Name>/childs/<child-device id>/```

Reply:
Wondering what's the benefit of that additional `childs` topic level is. The topic scheme like `tedge/<API Name>/<child-device id>` was already offering a way for selective subscription between the root device and child devices.